syntactic-whitelist/src/test/scala/WhitelistCheckerTests.scala

```diff
@@ -0,0 +1,446 @@
+import Features._
+import TestRunner.createTest
+import org.junit.{AfterClass, Test}
+
+import scala.meta.dialects
+
+class WhitelistCheckerTests {
+  import WhitelistCheckerTests._
+
+  @Test def allowLiteralsAndExpressions_should_allow_literals(): Unit = {
+    createTest(testController)
+      .onFile("LiteralsOnly")
+      .withFeatures(AllowLiteralsAndExpressions)
+      .withDialect(dialects.Sbt1)
+      .expectValid()
+  }
+
+  @Test def allowLiteralsAndExpressions_should_allow_literals_and_expressions(): Unit = {
+    createTest(testController)
+      .onFile("LiteralsAndExpressions")
+      .withFeatures(AllowLiteralsAndExpressions)
+      .withDialect(dialects.Sbt1)
+      .expectValid()
+  }
+
+  @Test def vals_should_be_rejected_when_not_allowed(): Unit = {
+    createTest(testController)
+      .onFile("Vals")
+      .withFeatures(ALL_FEATURES)
+      .exceptFeatures(AllowVals)
+      .expectInvalid(2 -> 1, 3 -> 1, 4 -> 1, 5 -> 1, 6 -> 1)
+  }
```

Comment:
Correct me if i'm wrong. The behavior is exactly the same as a blacklist checker which forbids `val`. It should be easy to construct a blacklist checker from a whitelist checker and vice versa if we know `ALL_FEATURES`.

Reply:
Interesting!

Additional comments:
- it seems that `exceptFeatures` is _always_ called after we call `withFeatures(ALL_FEATURES)`. Should do it in single method call?
- not very important, but I think we should remove the `Allow` prefix on features to have something that looks like the following:
  ~~~ scala
  .withoutFeatures(Vals)
  ~~~
  (instead of having “except” and “allow” on the same line)

Reply:
> Correct me if i'm wrong. The behavior is exactly the same as a blacklist checker which forbids `val`. It should be easy to construct a blacklist checker from a whitelist checker and vice versa if we know `ALL_FEATURES`.

Constructing a blacklist checker from the whitelist one should indeed be feasible as long as no new construct is added to the language, because if a construct is added `ALL_FEATURES` will not contain the whole language anymore. Constructing a whitelist checker from the blacklist one would require to define more rules in the blacklist checker, because the currently existing rules do not cover the whole language (but if I understood correctly it was not the goal).